### PR TITLE
Backport 2.7: Return internal error code, if failed 

### PR DESCRIPTION
**Pull request by RonEld**
## Description
Backport of #1096 to `mbedtls-2.7`

## Status
**READY**

### CODE REVIEW

File: library/ctr_drbg.c [hunk 1]
```diff
@@ -457,10 +456,11 @@ int mbedtls_ctr_drbg_random_with_add( void *p_rng,
 exit:
     mbedtls_zeroize( add_input, sizeof( add_input ) );
     mbedtls_zeroize( tmp, sizeof( tmp ) );
-    return( 0 );
+    return( ret );
 }
 
-int mbedtls_ctr_drbg_random( void *p_rng, unsigned char *output, size_t output_len )
+int mbedtls_ctr_drbg_random( void *p_rng, unsigned char *output,
+                             size_t output_len )
```
Comment (andresag01) [inline library/ctr_drbg.c]: I do not think this has anything to do with the change of the main PR, but its just style so I dont mind.

File: library/aes.c [hunk 2]
```diff
@@ -1031,13 +1058,18 @@ int mbedtls_aes_crypt_ctr( mbedtls_aes_context *ctx,
                        const unsigned char *input,
                        unsigned char *output )
 {
-    int c, i;
+    int c, i, ret = 0;
     size_t n = *nc_off;
+    unsigned char *out = output;
```
Comment (andresag01) [inline library/aes.c]: Is there a missing check here?
```
if( n > 0x0F )
```

Comment (RonEld) [reply library/aes.c]: I don't think so. `mbedtls-2.7` doesn't contain this check

File: library/ctr_drbg.c (hunk 1 above)
Comment (RonEld) [reply library/ctr_drbg.c]: Yes, it was a mistake in resolving the cherry-picking

**Comment by RonEld** [comment]
@andresag01 Thank you for the review.
I have addressed your comments

File: library/aes.c (hunk 2 above)
Comment (andresag01) [reply library/aes.c]: Ok, but should it be there?

Comment (RonEld) [reply library/aes.c]: Although I agree it's a good enhancement as parameter validation, it's not part of the scope of the PR

**Comment by mpg** [comment]
@andresag01 Please re-review. Both the main PR and the mbedtls-2.16 backport have been approved, but are held back by this backport's lack of approval.

**Comment by AndrzejKurek** [comment]
@andresag01 - please re-review this PR, as it is targeted at Mbed TLS 19.02.

**Comment by gilles-peskine-arm** [comment]
The 2.7 branch has reached its end of life, so I am closing this pull request.

**Pull request closed without merging**